FFCDemoPaymentService/Startup.cs

```diff
@@ -37,15 +40,23 @@ public void ConfigureServices(IServiceCollection services)
             services.AddScoped<IScheduleService, ScheduleService>();
             services.AddScoped<IPaymentService, PaymentService>();
             services.AddSingleton<IMessageAction<Schedule>, ScheduleAction>();
-            services.AddSingleton<IMessageAction<Payment>, PaymentAction>();
-            
+            services.AddSingleton<IMessageAction<Payment>, PaymentAction>(); 
+
             services.AddHealthChecks()
                 .AddCheck<ReadinessCheck>("ServiceReadinessCheck")
-                .AddCheck<LivenessCheck>("ServiceLivenessCheck");                            
+                .AddCheck<LivenessCheck>("ServiceLivenessCheck");
 
             services.AddControllers();
         }
 
+        private void AddTelemetry(IServiceCollection services)
+        {
+            string cloudRole = Environment.GetEnvironmentVariable("APPINSIGHTS_CLOUDROLE") ??
+                               "ffc-demo-payment-service-core";
```

Comment:
The convention for getting configuration values into your application is to use the `appsettings.json` file rather than directly calling the environment variables from your code.

There are a couple of reasons for this.

In order to satisfy dependency values, on Startup .Net Core runs through a set of steps to pull in config data.
1. resolves an instance of `IConfiguration` and passes to the `Startup` class.  It is this that provides the rest of these steps
2. reads `appsettings.json`
3. reads any environment specific json files overwriting values where matches, ie `appsettings.development.json`
4. reads any defined app secrets from connected secret stores, overwriting where values match
5. reads any environment variables, overwriting where values match
6. reads any command line arguments, overwriting where there is a match

So when the application starts up, it is already reading the environment variables and could have populated the value then.  If we use `GetEnvironmentVariable` then we are asking it to do something it's already done.

The other advantage of this approach is that you can configure your startup class to reload the values if the values file change which is useful during development but can also support hot reloading of prod services (if we ever need it).  Currently we're doing neither with this project but worth mentioning.

In order for .Net Core to be able to map environment variables correctly, they need to follow a convention.  So let's say you have an `appsettings.json` file which includes this value you want to be automatically replaced by an environment variable if it exists.
```
"AppInsightsCloudRole": "role"
```
If you call your environment variable "AppInsightsCloudRole" then the value `role` will automatically be replaced on app startup with the value of that environment variable.  

Here's the frustration, this convention is case sensitive.  So either you need to have a pascal case environment variable or you change the json file to block capitals, ie, `APPINSIGHTSCLOUDROLE:"role"`.  Both approaches frustrate me and as a lot of the other built in config such as logging needs to be pascal case, in this service we've opted for pascal case env vars as you can see from the other environment variables we've already used.  (gross I know)

Let's say you have a more complicated `appsettings.json` structure you wanted to satisfy such as:
```
"AppInsights": {
    "CloudRole": "ffc-demo-payment-service-core",
    "InstrumentationKey": "key"
  }
```
In order to override these two values you would need environment variables called `AppInsights__CloudRole` and `AppInsights__InstrumentationKey`.  Notice the double underscore, `__`.  Everytime you need to drop down a level in the JSON you add the double underscore and .Net Core will correctly override the values.

If you follow that convention then you'll be able to use the `Configuration` class in `Startup` anytime you want to reference a configuration property regardless of whether it originated in the file, env var or a secret store.

You have the option of reading a single value from the config or deserialize a whole section as a class.  You can add classes to the .Net Core DI container so they're automatically satisfied by consuming classes.  You'll see this was done for the messaging.

So my recommendation would be to make the following changes.

`appsettings.json`
Add the following section:
```
"AppInsights": {
    "CloudRole": "ffc-demo-payment-service-core",
    "InstrumentationKey": "key"
  }
```
`docker-compose.yaml`
Change the environment variables to match the convention needed for .Net Core to override them.
```
environment:
      AppInsights__InstrumentationKey: ${APPINSIGHTS_INSTRUMENTATIONKEY}
      AppInsights__CloudRole: ffc-demo-payment-service-core-local
```
`Startup.cs`
In the `AddTelemetry` method, replace the `GetEnvironmentVariable` call with:
```
string cloudRole = Configuration.GetValue<string>("AppInsights:CloudRole");
```
You can also do this syntax if you prefer:
```
string cloudRole = Configuration.GetSection("AppInsights").GetSection("CloudRole").Value
```
I prefer the former for single values and only reference full sections when deserializing to classes.  But that's personal preference, whatever you think reads best.

Reply:
thanks, I've made the changes, I've used `ApplicationInsights` as the appsetting section name as the key is automatically picked up with that naming convention.